### PR TITLE
Feature/refactor and type expectation diagnostic report

### PR DESCRIPTION
**Pull request by abegong**
WIP. Not yet ready for review.

### CODE REVIEW

File: great_expectations/core/expectation_diagnostics/expectation_diagnostics.py [hunk 1]
```diff
@@ -0,0 +1,251 @@
+from dataclasses import dataclass
+from typing import List, Tuple
+
+from great_expectations.core.expectation_diagnostics.expectation_test_data_cases import (
+    ExpectationTestDataCases,
+)
+from great_expectations.core.expectation_diagnostics.supporting_types import (
```
Comment (donaldheppner) [inline great_expectations/core/expectation_diagnostics/expectation_diagnostics.py:7]: Remove the supporting_types package

File: great_expectations/core/expectation_diagnostics/expectation_diagnostics.py [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+    ExpectationDescriptionDiagnostics,
+    ExpectationDiagnosticCheckMessage,
+    ExpectationDiagnosticMaturityMessages,
+    ExpectationErrorDiagnostics,
+    ExpectationExecutionEngineDiagnostics,
+    ExpectationMetricDiagnostics,
+    ExpectationRendererDiagnostics,
+    ExpectationTestDiagnostics,
+)
+from great_expectations.types import SerializableDictDot
```
Comment (donaldheppner) [inline great_expectations/core/expectation_diagnostics/expectation_diagnostics.py:19]: modules should be semantic, not reflecting the nature of the code in the module

File: great_expectations/core/expectation_diagnostics/expectation_diagnostics.py [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+        return str_
+
+    @staticmethod
+    def _check_library_metadata(
+        library_metadata: AugmentedLibraryMetadata,
+    ) -> ExpectationDiagnosticCheckMessage:
+        """Check whether the Expectation has a library_metadata object"""
+
+        return ExpectationDiagnosticCheckMessage(
+            **{
```
Comment (donaldheppner) [inline great_expectations/core/expectation_diagnostics/expectation_diagnostics.py]: use dot notation

File: great_expectations/core/expectation_diagnostics/expectation_diagnostics.py [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+
+        message = "Has a docstring, including a one-line short description"
+        if "short_description" in description:
+            short_description = description["short_description"]
+        else:
+            short_description = None
+        if short_description not in {"", "\n", "TODO: Add a docstring here", None}:
+            return ExpectationDiagnosticCheckMessage(
+                **{
+                    "message": message,
```
Comment (donaldheppner) [inline great_expectations/core/expectation_diagnostics/expectation_diagnostics.py]: use dot notation

File: tests/expectations/test_expectation_diagnostics.py [hunk 5]
```diff
@@ -0,0 +1,430 @@
...
+    ) == ExpectationDiagnosticCheckMessage(
+        **{
+            "message": "Has at least one positive and negative example case, and all test cases pass",
+            "passed": False,
+        }
+    )
+
+
+def test__check_example_cases__with_enough_test_cases_and_no_failing():
+    print(expectation_test_data_case)
```
Comment (donaldheppner) [inline tests/expectations/test_expectation_diagnostics.py]: please remove

**Review by donaldheppner — CHANGES_REQUESTED** [review]

File: tests/expectations/test_expectation_diagnostics.py (hunk 5 above)
Comment (abegong) [reply tests/expectations/test_expectation_diagnostics.py]:
```suggestion
```

**Pull request merged**